### DISCUSSION TITLE
Cloak Use Case Ideas

### DISCUSSION THREAD

**Discussion by sipsma** [dagger/cloak#143] [Ideas]
Starting this as a way of collecting ideas for cool use cases and things to try out with cloak. Real-world use cases are especially great, but also feel free to throw out "crazy ideas" too.

**Comment by sipsma** [comment]
An extension that uses infrastructure-as-code tools like Pulumi. Could be a general purpose "Pulumi extension" or more like an example of an extension for your app's stack that uses Pulumi by just importing and executing code.

**Reply by marcosnils** [reply]
:eyes: @rawkode

**Comment by sipsma** [comment]
Somewhat crazy idea (but may be something to it): A `coreutils` extension, something along the lines of:
```
extend type Filesystem {
   cat(path: String!) String!
   grep(path: String!, searchPattern: String!, ...) [String!]
   sed(...)
   <etc.>
}
```
(we probably want input types rather than direct args, but you get the idea)

This would be implemented by actually running the corresponding binaries on the filesystem.

This would be fun+cool by itself (can get access to these powerful+familiar command line tools from any language interface), but there may be a more interesting generic idea here if you expand out from just "coreutils".

**Reply by marcosnils** [reply]
> This would be fun+cool by itself (can get access to these powerful+familiar command line tools from any language interface), but there may be a more interesting generic idea here if you expand out from just "coreutils".

loved it. `busybox` is coming to Dagger :D

**Comment by sipsma** [comment]
Crazy idea: An extension for invoking cloud function-as-a-service deployments like AWS Lambda. Rough example:
```graphql
{
  aws(accountid: "blah", secretID: "theID", ...) {
    lambda(arn: "the:arn", input: {<json blob>})
  }
}
```

With an SDK that can introspect a Lambda deployment and generate schemas from it, you could end up with a typed version of the above.

**Comment by sipsma** [comment]
Kubernetes comes up a lot (also exists in dagger europa's universe today) (most basic thing would be to deploy to it obviously, but potentially could involve any part of the vast k8s api).

Not strictly necessary to use, but there are graphql endpoints for k8s today which may be worth looking into: https://github.com/onelittlenightmusic/kubernetes-graphql

**Comment by sipsma** [comment]
Define a development environment as an extension. I want to define my devenv in my own github repo and then start it by just importing the extension and executing it with cloak (probably just the cloak cli will suffice for this case).

This will require at minimum TTY support (as mentioned here: https://github.com/dagger/cloak/issues/125) and probably need support for some sort of continuous local dir syncing (I know the devs of the HLB project have used mutagen over sockets mounted into buildkit containers for similar things in the past).

**Comment by sipsma** [comment]
apko integration could be very interesting: https://github.com/chainguard-dev/apko

Could just shell out to apko, use the underlying go apk library they are working on to directly recreate the same idea, or possibly update apko to be usable as a library.

**Comment by marcosnils** [comment]
Thinking about use-cases for rapid adoption / usefulness. 

- cosign (https://github.com/sigstore/cosign) integration to allow image signing independently on what's happening on buildkit
- most popular cloud providers CLI integrations (since adding resource specific API's will probably take time). Maybe the API for AWS could be something like
```graphql 
{
  aws(accountid: "blah", secretID: "theID", ...) {
    # Map could be a gqlgen (or equivalent framework) scalar Map type (https://gqlgen.com/reference/scalars/#map) 
    cli(args: ["s3api", "list-buckets"]) Map!  
   }
}
```

**Comment by dolanor** [comment]
Makefile -> Buildkit translator.

Each target in a Makefile becomes a step in cloak. The dependency tree would be around the same, but we could have a more powerful caching model (than the file one that Makefile uses), that would allow us to setup some dependency first before executing some task.
